runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -4788,341 +4800,294 @@ J9::Z::TreeEvaluator::anewArrayEvaluator(TR::Node * node, TR::CodeGenerator * cg
-      // Discontiguous array layout: |  class  | mustBeZero |  size   | ...
-      // Contiguous array layout:    |  class  |    size    | ...
-      cursor = generateRXInstruction(cg, TR::InstOpCode::ST, node, firstDimLenReg, generateS390MemoryReference(targetReg, fej9->getOffsetOfContiguousArraySizeField()/* = "mustBeZero" offset!*/, cg));
-      iComment("Init 1st dim mustBeZero field.");
-      cursor = generateRXInstruction(cg, TR::InstOpCode::ST, node, firstDimLenReg, generateS390MemoryReference(targetReg, fej9->getOffsetOfDiscontiguousArraySizeField(), cg));
-      iComment("Init 1st dim size field.");
-   if (TR::Compiler->om.compressObjectReferences())
+      cursor = generateRSInstruction(cg, TR::InstOpCode::SLA, node, dim2SizeReg, trailingZeroes(componentSize), cursor);
+      // Only a positive int32 with no overflow is acceptable. This will guarantee no overflow in multiplication instruction.
+      cursor = generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BRNP, node, inlineAllocFailLabel, cursor);
+      }
+   int32_t headerSize= TR::Compiler->om.contiguousArrayHeaderSizeInBytes();
+   bool alignSecondDim = !(OMR::aligned(headerSize, alignmentConstant) && OMR::aligned(componentSize, alignmentConstant));
+   cursor = generateRIInstruction(cg, TR::InstOpCode::AGHI, node, dim2SizeReg, alignSecondDim ? (headerSize + alignmentConstant - 1) : headerSize, cursor);
+   if (alignSecondDim)
+      cursor = generateRILInstruction(cg, TR::InstOpCode::NILF, node, dim2SizeReg, -alignmentConstant, cursor);
+   cursor = generateRRInstruction(cg, TR::InstOpCode::LGR, node, sizeReg, dim2SizeReg, cursor);
+
+   cursor = generateS390LabelInstruction(cg, TR::InstOpCode::label, node, zeroSecondDimLabel, cursor);
+   // If jumping from zero second dim match, set second dim size to discontiguous. Otherwise this wont change values.
+   cursor = generateRRInstruction(cg, TR::InstOpCode::LR, node, dim2SizeReg, sizeReg, cursor);
+   // Total size = (second dim size * number of second dim arrays) + first dim size.
+   // Max possible value is uint32.max * int32.max which will not overflow.
+   cursor = generateRREInstruction(cg, TR::InstOpCode::MSGR, node, sizeReg, dim1SizeReg, cursor);
+
+   // Size = (element size * dimension length) + header size. final size must get aligned.
+   cursor = generateRSInstruction(cg, TR::InstOpCode::SLLG, node, dim1SizeReg, dim1SizeReg, trailingZeroes(elementSize), cursor);
+   bool alignFirstDim = !(OMR::aligned(headerSize, alignmentConstant) && OMR::aligned(elementSize, alignmentConstant));
+   cursor = generateRIInstruction(cg, TR::InstOpCode::AGHI, node, dim1SizeReg, alignFirstDim ? (headerSize + alignmentConstant - 1) : headerSize, cursor);
+   if (alignFirstDim)
+      cursor = generateRILInstruction(cg, TR::InstOpCode::NILF, node, dim1SizeReg, -alignmentConstant, cursor);
+
+   cursor = generateRREInstruction(cg, TR::InstOpCode::AGR, node, sizeReg, dim1SizeReg, cursor);
```

Comment:
If we reach here, do we expect to be dealing with negative values ? What would be the consequences of using `ALGR`?

Reply:
So can you use ALGR? that would guarantee to not cause fixed point overflow.